### PR TITLE
(FACT-2555)Create OS hierarchy and mechanism for loading it

### PR DESCRIPTION
**Pull request by BogdanIrimie**

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -90,6 +90,7 @@ RSpec/FilePath:
     - 'spec/framework/formatters/legacy_fact_formatter_spec.rb'
     - 'spec/framework/formatters/yaml_fact_formatter_spec.rb'
     - 'spec/framework/utils/utils_spec.rb'
+    - 'spec/framework/detector/os_hierarchy_spec.rb'
```
[the remark this answers is not in the document]
Comment (IrimieBogdan) [reply .rubocop_todo.yml:93]: Created ticket https://tickets.puppetlabs.com/browse/FACT-2583

File: spec/framework/detector/os_detector_spec.rb [hunk 2]
```diff
@@ -3,68 +3,178 @@
...
 
-    it 'calls Facter::Resolvers::OsRelease with version' do
-      OsDetector.instance
-      expect(Facter::Resolvers::OsRelease).to have_received(:resolve).with(:version)
+    context 'when os cannot be detected' do
+      before do
+        RbConfig::CONFIG['host_os'] = 'my_custom_os'
+      end
+
+      it 'raise error if it could not detect os' do
```
Comment (Filipovici-Andrei) [inline spec/framework/detector/os_detector_spec.rb:100]: an extra s at raise

File: spec/framework/detector/os_detector_spec.rb [hunk 3]
```diff
@@ -3,68 +3,178 @@
...
+  let(:os_hierarchy) { instance_spy(Facter::OsHierarchy) }
+
   before do
     Singleton.__init__(OsDetector)
-  end
 
-  it 'detects os as macosx' do
-    RbConfig::CONFIG['host_os'] = 'darwin'
-    expect(OsDetector.instance.identifier).to eq(:macosx)
+    allow(Facter::OsHierarchy).to receive(:new).and_return(os_hierarchy)
```
Comment (Filipovici-Andrei) [inline spec/framework/detector/os_detector_spec.rb:11]: In my opinion as a general practice, I think that you shouldn't mock application code that is not slow. We should mock only system calls.

**Review by Filipovici-Andrei — CHANGES_REQUESTED** [review]

File: spec/framework/detector/os_hierarchy_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  before do
+    allow(Facter::Util::FileHelper)
+      .to receive(:safe_read)
+      .with('os_hierarchy.json')
+      .and_return(load_fixture('os_hierarchy').read)
+  end
+
+  describe '#construct_hierarchy' do
+    context 'when searched_os is ubuntu' do
```
[the remark this answers is not in the document]
Comment (IrimieBogdan) [reply spec/framework/detector/os_hierarchy_spec.rb:14]: Added tests for intermediary and root nodes

File: spec/framework/detector/os_hierarchy_spec.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+        expect(hierarchy).to eq(%w[Linux Debian Ubuntu])
+      end
+    end
+
+    context 'when searched_os is nil' do
+      it 'constructs hierarchy' do
+        hierarchy = os_hierarchy.construct_hierarchy(nil)
+
+        expect(hierarchy).to eq(%w[])
```
[the remark this answers is not in the document]
Comment (IrimieBogdan) [reply spec/framework/detector/os_hierarchy_spec.rb:42]: Removed %w

File: spec/framework/detector/os_detector_spec.rb (hunk 2 above)
Comment (IrimieBogdan) [reply spec/framework/detector/os_detector_spec.rb:100]: Added extra `s`

File: spec/framework/detector/os_detector_spec.rb (hunk 3 above)
Comment (IrimieBogdan) [reply spec/framework/detector/os_detector_spec.rb:11]: As this is a unit test, we should mock everything that is outside the unit we are testing, in this case the unit under test is `Facter::OsHierarchy`
If we don't do this, the unit tests will become integration tests.

File: lib/framework/detector/os_hierarchy.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+module Facter
+  class OsHierarchy
+    @searched_path = []
+
+    def initialize
+      json_file = Util::FileHelper.safe_read('os_hierarchy.json')
+      @json_data = JSON.parse(json_file)
```
[the remark this answers is not in the document]
Comment (IrimieBogdan) [reply lib/framework/detector/os_hierarchy.rb:7]: I added error handling for it, but this file is placed inside the gemfile and is like an internal config for facter, it should not be edited by users.

[the remark this answers is not in the document]
Comment (gimmyxd) [reply lib/framework/detector/os_hierarchy.rb:8]: Ok, i taught the users can jump in and add new os-es on the fly :D

**Review by Filipovici-Andrei — DISMISSED** [review]

File: spec/framework/detector/os_detector_spec.rb [hunk 7]
```diff
@@ -3,68 +3,178 @@
...
+      it 'detects os as macosx' do
+        expect(OsDetector.instance.identifier).to eq(:macosx)
+      end
 
-  it 'detects os as aix' do
-    RbConfig::CONFIG['host_os'] = 'aix'
-    expect(OsDetector.instance.identifier).to eq(:aix)
-  end
+      it 'calls hierarchy construction with macosx identifier' do
+        expect(os_hierarchy).to have_received(:construct_hierarchy).with(:macosx)
```
[the remark this answers is not in the document]
Comment (IrimieBogdan) [reply spec/framework/detector/os_detector_spec.rb:28]: Added `OsDetect.instance` to all context that were missing it.

**Review by gimmyxd — APPROVED** [review]

**Pull request merged**